fdk-aac encoder + libshout-idjc + vcpkg Windows dependencies integration

Be-ing opened the pull request:
This is a rebased combination of #3543 and #3594 which enables AAC encoding on Windows, macOS, and Linux.

Commit:
fdk-aac encoder

daschuer committed:
Renamed libshout folder to libshout-idjc

daschuer committed:
update to libshout-idjc-2.4.1

daschuer committed:
fix loading of libfdk-aac if no symlink without a version is provided

daschuer committed:
fix line break

daschuer committed:
fix include files

daschuer committed:
Enable internal libshout-idjc for all targets

daschuer committed:
clazy fixes

daschuer committed:
Fix indent

daschuer committed:
Don't use config.h configure by cmake instead, taken from https://github.com/Holzhaus/icecast-libshout

daschuer committed:
fix os.h

daschuer committed:
Remove unsused strings.h

daschuer committed:
replace illegal void* arythmetric

daschuer committed:
Try loading libfdk-aac.so.2 first

daschuer committed:
Avoid possible string truncation.

daschuer committed:
Avoid extra buffer for sample conversion

daschuer committed:
link libshout-idjc against pthread

daschuer committed:
Link against pthread on windows

Holzhaus committed:
CMake: Rewrite Ogg/Vorbis find module

Holzhaus committed:
CMake: Allow building if FFMPEG version is missing

daschuer committed:
Fix file extensions during recording.

Be-ing committed:
use Windows dependencies from vcpkg

daschuer committed:
Merge remote-tracking branch 'upstream/2.3' into fdkaac-encoder3

daschuer committed:
Fix linking to system installed libshout-idjc

daschuer committed:
Merge remote-tracking branch 'upstream/2.3' into fdkaac-encoder3

daschuer committed:
Re-enabled Quality slider for AAC recording

daschuer committed:
Avoid log spam when probing fdkaac libraries

daschuer committed:
Check return value from aacEncoder_SetParam

daschuer committed:
Check for SBR support

daschuer committed:
Replace QString::null

daschuer committed:
Improve error handling when initalize the encoders for recording or broadcasting

daschuer committed:
Prevent double message box in case of recording issues.

daschuer committed:
Fix second pop up in case of encoder issues during broadcasting.

daschuer committed:
fixed shadowed varibale issue

daschuer committed:
removed stray ;

daschuer committed:
Fix redundant message box when fdk-aac is not installed.

daschuer committed:
Update error message.

Co-authored-by: Be <be.0@gmx.com>

Be-ing committed:
use macOS build environment with fdk-aac fork without HE-AAC

daschuer committed:
replace isNull() with isEmpty()

Co-authored-by: Be <be.0@gmx.com>

daschuer committed:
Use more QStringLiteral

daschuer committed:
remove useless comment

daschuer committed:
Make the encoding keys more readable.

Co-authored-by: Be <be.0@gmx.com>

Be-ing committed:
CMake: copy libfdk-aac into Windows and macOS packages

Be-ing committed:
EncoderFdkAac: change m_library to a std::unique_ptr

Be-ing committed:
EncoderFdkAac: prefer external libraries that may support HE-AAC

and search for library in bundle on macOS

Be-ing committed:
CMake libshout-idjc: link pthreads4w from vcpkg

Be-ing committed:
EncoderFdkAac: update name of BUTT DLL

daschuer commented:
Lets go in steps an merge https://github.com/mixxxdj/mixxx/pull/3543 first. 
There is no need to create a giant PR.

Be-ing commented:
>  There is no need to create a giant PR.

Yes, it all needs to be tested together to ensure it works on every OS.

daschuer commented on CMakeLists.txt:
```diff
@@ -2219,29 +2224,49 @@ if(FAAD)
+# so copy it into the Windows and macOS packages, but do not link to it.
+find_library(FDK_AAC_LIBRARY fdk-aac)
+# On Windows find_library finds the .lib file, but the installer needs the .dll file.
+find_file(FDK_AAC_DLL fdk-aac.dll PATH_SUFFIXES ${CMAKE_INSTALL_BINDIR})
```
```suggestion
find_file(FDK_AAC_DLL libfdk-aac-2.dll PATH_SUFFIXES bin)
```
This should work. 
see: 
https://github.com/mixxxdj/mixxx/pull/3543/commits/90033594106a2b4f383b06e006cacf058b5e6596

Be-ing replied:
This already works. I'm not going to break it with a less sustainable implementation.

daschuer replied:
How ${CMAKE_INSTALL_BINDIR} is "."

daschuer replied:
Ah I see CMAKE_INSTALL_BINDIR i got confused with MIXXX_INSTALL_BINDIR  which is "."

daschuer replied:
The issue is here that it now installs a fdk-aac.dll with an unusual name. The usual name is libfdk-aac-2.dll

Be-ing replied:
That is incorrect. DLLs do not conventionally have a `lib` prefix nor a version suffix on Windows. There is nothing to fix here. It works.

daschuer replied:
Yes I know, but we are the first following this convention.

Be-ing committed:
CMake: cleanup hacks for legacy Windows build environment

Be-ing committed:
CMake: add user feedback for FindFFMPEG.cmake module

Be-ing committed:
CMake: enable FFMPEG if it is found

Be-ing commented:
All tests on Windows failed with "Exit code 0xc0000135"???

uklotzde commented:
We now have 2 competing PRs with libshout-idjc. I am confused.

daschuer commented:
Let's go in steps and merge this smaller PR first.
It works with the current windows build environment and is tested with that.

Shipping of fdk-aac-free is a separate topic.

Be-ing commented:
There is no point merging a PR that introduces a feature that can't be tested. This PR is tested on Windows, macOS, and Linux and it AAC encoding works.

daschuer commented:
Oh sorry, I have commented the wrong PR. This should go to.
https://github.com/mixxxdj/mixxx/pull/3543

> There is no point merging a PR that introduces a feature that can't be tested.

Isn't this a pintless discussion?

This PR is a follow up of mine. It piles up many more changed, replacing other libraries.
The internal libshout-idjc is not necessarily to merge together with them.

My branch is tested by me and others, I don't understand why you keep thinking that it is not tested.

It was you request to package the fdk-aac-free version.
This is now in this extra branch which is not reviewed and tested by a third party.

Be-ing commented:
No, you have introduced conflicting untested commits in your branch. I cannot just rebase my work if that branch is merged.

daschuer commented:
Just merge and resolve the conflicts as usual.
Every commit is tested as far as possible.

Be-ing commented:
> All tests on Windows failed with "Exit code 0xc0000135"???

Another new build environment failed. I think there is a regression in upstream vcpkg??

Be-ing commented:
I suppose we could merge this using the environment from the fdk_aac_fedora vcpkg branch. But I'd like to get a build from the 2.3 branch working...

Be-ing commented:
I have no idea what's causing this "Exit code 0xc0000135" error. I tried reverting the only recent commit in vcpkg that seemed potentially relevant but it made no difference. I guess I'll open a bug report upstream?

Be-ing commented:
> All tests on Windows failed with "Exit code 0xc0000135"??? But only with the new build environments from the 2.3 branch of the vcpkg repo... the environment from the fdk_aac_fedora branch works...

https://github.com/microsoft/vcpkg/issues/16078

Be-ing commented:
Okay https://github.com/mixxxdj/vcpkg 2.3 branch is in a working state (including caching on GitHub Actions). I do not know why merging microsoft/vcpkg master after that breaks the tests with exit code 0xc0000135. I will continue debugging that on another branch, but there's no need to hold off merging this because of that issue.

Be-ing commented:
It looks like there is a regression in libsndfile?
```
2021-02-09T02:03:29.8096118Z Info []: Opened file "D:/a/mixxx/mixxx/src/test/id3-test-data/cover-test.ogg" using provider "libsndfile"
2021-02-09T02:03:29.8097510Z ..\src\test\soundproxy_test.cpp(121): error: The difference between expected[i] and actual[i] is 0.62643324583768845, which exceeds kMaxDecodingError, where
2021-02-09T02:03:29.8099228Z expected[i] evaluates to -0.042898960411548615,
2021-02-09T02:03:29.8099784Z actual[i] evaluates to -0.66933220624923706, and
2021-02-09T02:03:29.8100399Z kMaxDecodingError evaluates to 0.0099999997764825821.
2021-02-09T02:03:29.8101022Z Decoding mismatch after seeking backward
```
Why is libsndfile being used for an Ogg file??

Be-ing commented:
Yup, there was a regression in libsndfile. Reverting to 1.0.30...

Be-ing committed:
factor out SemanticVersion into a utility class

Be-ing committed:
SoundSourceSndFile: do not support ogg for libsndfile >= 1.0.30

https://github.com/libsndfile/libsndfile/issues/643 was not fixed in libsndfile
1.0.31.

Be-ing commented:
I tripled checked and AAC encoding is working on Windows and macOS. Please review.

Be-ing commented:
We could get rid of the QLibrary and just link fdk-aac dynamically and tell users who want HE-AAC to replace the DLL/dylib we ship with Mixxx. But let's save that for another PR.

Be-ing commented:
> We could get rid of the QLibrary and just link fdk-aac dynamically and tell users who want HE-AAC to replace the DLL/dylib we ship with Mixxx. But let's save that for another PR.

I just went ahead and did this. It got rid of a lot of complexity. If users want to use the fdk-aac DLL/dylib from BUTT to use HE-AAC, they can replace the DLL/dylib shipped with Mixxx. There's no need for the hacky searching for the DLL in a BUTT installation. That's cumbersome anyway because BUTT ships the DLL/dylib separate from BUTT.

Be-ing committed:
Encoder: fix build with OPUS=OFF

Be-ing committed:
EncoderFdkAac: use #pragma once

Be-ing committed:
avoid symbol collision with fdk-aac

This is just a quick hack. We should move all Mixxx code into the mixxx
namespace.

Be-ing commented:
I stumbled on an annoying symbol collision with fdk-aac. It is fixed now, but we should put more work into moving all Mixxx code into the `mixxx` namespace.

Be-ing commented:
I tested again on Windows and macOS now that fdk-aac is linked to Mixxx. It still works. Replacing the DLL/dylib from Mixxx with the one from http://danielnoethen.de/butt/manual.html#_install_aac_libraries successfully enables HE-AAC.

uklotzde commented:
@Be-ing  @daschuer How much has this PR diverged from #3543?

Be-ing commented:
To clarify, I started working with an integration branch on top of #3543 but @daschuer cherry picked commits that weren't ready as I was still testing them, then made more changes I disagreed with. That issue we disagreed over is now irrelevant because this branch links fdk-aac like any other dynamic library instead of using QLibrary.

Be-ing commented:
Should we add libfdk-aac-dev as a dependency for Debian? It is in Debian's nonfree repository. I don't particularly care either way.

daschuer commented:
>> We could get rid of the QLibrary and just link fdk-aac dynamically and tell users who want HE-AAC to replace the DLL/dylib we ship with Mixxx. But let's save that for another PR.

> I just went ahead and did this. It got rid of a lot of complexity. If users want to use the fdk-aac DLL/dylib from BUTT to use HE-AAC, they can replace the DLL/dylib shipped with Mixxx. There's no need for the hacky searching for the DLL in a BUTT installation. That's cumbersome anyway because BUTT ships the DLL/dylib separate from BUTT.

This is a bead idea. Because If anything goes wrong Mixxx will crash during start with some obscure error messages. 

Using the DLL directly from the BUTT install folder is mainly for older installation of BUTT where the Library was part of a single instillation package. It is also convenience if BUTT is installed and a user just want to check out BUTT.

Be-ing commented:
> This is a bead idea. Because If anything goes wrong Mixxx will crash during start with some obscure error messages.

I tested this. The error message is not obscure, it says exactly what is wrong: `fdk-aac.dll could not be found`. As long as we have clear documentation I think it will be fine. I tested on both macOS and Windows, it is quite simple. Probably few users will care about using HE-AAC anyway.

> Using the DLL directly from the BUTT install folder is mainly for older installation of BUTT where the Library was part of a single instillation package. It is also convenience if BUTT is installed and a user just want to check out BUTT.

Thank you for explaining the context. Now that BUTT does not ship fdk-aac together with BUTT, I think searching for the library in the BUTT installation is obsolete. We should not assume any users have BUTT installed or have even heard of BUTT. Putting fdk-aac in the BUTT installation now requires them to install two extra downloads. If we tell them to just replace the library shipped with Mixxx with the one BUTT hosts, that is only one file to download.

daschuer commented:
@uklotzde 
This branch is a combination of the libfdk-aac integration and the change of the windows build environment which updates many other libraries. I would prefer to go in smaller steps. Adding the libfdk-aac free as well was an unnecessary feature creep.

daschuer commented:
> I tested this. The error message is not obscure, it says exactly what is wrong: fdk-aac.dll could not be found. As long as we have clear documentation I think it will be fine. I tested on both macOS and Windows, it is quite simple. Probably few users will care about using HE-AAC anyway.

The other issue is that we than need to require the libfdk-aac on Linux. The user will have hard times to install the non free version because both packages are conflicting. 
In Linux Mixxx will silently crash without the libfdk-aac found. 

> Thank you for explaining the context. Now that BUTT does not ship fdk-aac together with BUTT, I think searching for the library in the BUTT installation is obsolete. We should not assume any users have BUTT installed or have even heard of BUTT. 

Why not? I think a normal broadcasting user will likely have BUTT before Mixxx. These users will have zero hassle, and no extra download at all. HE-AAC will just work.

Be-ing commented:
> The other issue is that we than need to require the libfdk-aac on Linux. The user will have hard times to install the non free version because both packages are conflicting.

Hmm, you have a point there. Using QLibrary allows the user to choose which fdk-aac package to use without needing to cause trouble for packagers or making users compile Mixxx on their own.

Be-ing commented:
I have rebased to remove the linking and go back to using QLibrary. I have backed up the QLibrary commit on [this branch](https://github.com/Be-ing/mixxx/tree/fdk-aac_link) in case we change our decision in the future.

JoergAtGithub commented:
I did a short test of the Windows build artifact and it works well. But I didn't had the possibilty to test the streaming functionality.

JosepMaJAZ commented:
I also did a quick test.

Played several songs with the controller, tested different waveforms, tested recording to aac and broadcasting to AAC.

About broadcasting to AAC, it works with icecast 2, but it cannot be used with shoutcast due to a missing change in shoutconnection.cpp. 
After determining the format, it determines the protocol, and there, we have a check that shoutcast can only use AAC (because it does not support vorbis or opus), but it does support AAC. (or should).
Broacasting to MP3 also works.

It has the known problem with libmodplug where songs might not play correctly due to multithreading issues.

I supposedly tested libkeyfinder (I changed the setting in the preferences, accepted and analysed some tracks.

I didn't test LV2 since I don't have any.

Installation folder structure seems correct, but I saw "fdk-aac.dll" and "libfdk-aac-2.dll"

Btw... This MIGHT have fixed the issues that I was having with WDM-KS lately with Windows (and needing to use WASAPI)

Be-ing commented:
> About broadcasting to AAC, it works with icecast 2, but it cannot be used with shoutcast due to a missing change in shoutconnection.cpp.
After determining the format, it determines the protocol, and there, we have a check that shoutcast can only use AAC (because it does not support vorbis or opus), but it does support AAC. (or should).

I don't know what would be required. Could you make a PR for my fork implementing this?

> I didn't test LV2 since I don't have any.

https://github.com/DISTRHO/PawPaw/releases/tag/v1.0

> Btw... This MIGHT have fixed the issues that I was having with WDM-KS lately with Windows (and needing to use WASAPI)

That would be great. It is using [my PortAudio branch](https://github.com/PortAudio/portaudio/pull/461) which is based on PortAudio's master branch.

JosepMaJAZ commented:
About the broadcasting, I mean this line of the code:
https://github.com/mixxxdj/mixxx/pull/3615/files#diff-11bde85373fd2b4777f42d158c6faf2376fae587a1e5d5c617ff14d733650c06L432

We are only allowing MP3 with shoutcast, but now we should allow AAC too. 


Ok, Downloaded that LV2 plugins installer and I tested one of them ( MDA loudness. hehe.. I used to have several of the MDA ones in VST2 form)

Be-ing commented:
So should that `if` statement just be removed?

Be-ing committed:
ShoutConnection: allow AAC for Shoutcast

Be-ing commented:
@JosepMaJAZ please review the commit I just pushed. I don't really know what I'm doing there.

JosepMaJAZ commented:
Thanks.

That worked, now AAC can be used with both, Icecast and shoutcast. 
The only missing thing would be a TLS ( https) test, but I don't currently have the means to do that.

On the recording side, we still miss encapsulating the aac stream into an mp4/m4a container, which should be easy now that we integrate ffmpeg, but I haven't investigated it yet (that can be worked on under linux as well)

Be-ing commented:
> On the recording side, we still miss encapsulating the aac stream into an mp4/m4a container, which should be easy now that we integrate ffmpeg, but I haven't investigated it yet (that can be worked on under linux as well)

This isn't strictly necessary. While testing this I noticed that SoundSourceMediaFoundation and SoundSourceCoreAudio were not enabled for `.aac` files but I fixed those (#3616, #3614).

Be-ing commented:
I think this is ready to merge.

Be-ing commented:
I rebased to remove enabling KeyFinder given the situation with #3476...

Be-ing commented:
Merge? This needs testing which it won't get much more of sitting unmerged.

daschuer commented on src/encoder/encoderfdkaac.cpp:
```diff
@@ -0,0 +1,437 @@
+            QStringLiteral("/../Frameworks/libfdk-aac"));
+    libnames << bundlePath.absoluteFilePath();
+#endif
+    libnames << QStringLiteral("fdk-aac");
```
This code does not work for me. It obfuscates which library is really loaded, Because it relies on target specific code here and in the Qt library. Only if you know both you understand what happen. 
The solution in https://github.com/mixxxdj/mixxx/pull/3543 works much better, readable like a book:
https://github.com/mixxxdj/mixxx/blob/b660eb23165bf04b30f2d9c133758c35f38d2b90/src/encoder/encoderfdkaac.cpp#L66

Be-ing replied:
Can we please stop quibbling over this inconsequential implementation detail and get a release out?

Be-ing replied:
I still think reimplementing abstractions that Qt provides for us is a bad idea.

daschuer replied:
The Qt abstraction is made for the case Qt considers as the regular case. But we are not at the regular case here and we need to hack around it. The proposed code is only readable if you full understand the QT abstraction. I had to read the Qt source to fully understand it. In the original implementation we build a easy to understand string list, tweak-able per target.

And by the way, if this is so quibbling for you, just revert it and we can make progress.

Be-ing replied:
I have tested this on 3 OSes. It works. Qt documentation explains how it works. I am not putting any more effort into this laborious process of testing that everything works across 3 OSes with an alternative implementation reimplementing what Qt already does for us.

daschuer replied:
The original version works as well. Now we have cluttered hard to maintain version that relies on Qt implementation details.

daschuer commented on CMakeLists.txt:
```diff
@@ -1350,6 +1352,25 @@ if(MSVC)
+    DIRECTORY "${QT_TRANSLATIONS}"
+    DESTINATION "${MIXXX_INSTALL_DATADIR}"
+    # QT 5 translations have been separated into several files, and most of the qt_xx.qm files
+    # contain just shortcuts to load the qtbase, qtmultimedia etc files.
```
Please add a conclusion that leads to the regex.

Be-ing replied:
What regex? There is no regex here...

Be-ing replied:
The regex is unchanged from before...

daschuer commented on CMakeLists.txt:
```diff
@@ -1350,6 +1352,25 @@ if(MSVC)
 endif()
 
+if(WIN32)
+  find_file(QT_TRANSLATION_FILE qt_de.qm PATH_SUFFIXES share/qt5/translations)
```
Please add a comment what this block is about. Does it install QT's translations?

Be-ing replied:
obviously...

Be-ing replied:
I'll add a comment. `qt_de.qm` is just one arbitrary file in that directory.

daschuer commented on CMakeLists.txt:
```diff
@@ -2219,29 +2217,49 @@ if(FAAD)
+# FDK-AAC is loaded dynamically at runtime by EncoderFdkAac using QLibrary,
+# so copy it into the Windows and macOS packages, but do not link to it.
+find_library(FDK_AAC_LIBRARY fdk-aac)
+# On Windows find_library finds the .lib file, but the installer needs the .dll file.
```
Why we search for the lib in the fist place?

Be-ing replied:
The comments explain that clearly...

daschuer commented on CMakeLists.txt:
```diff
@@ -2219,29 +2217,49 @@ if(FAAD)
+  message(STATUS "Found fdk-aac DLL: ${FDK_AAC_DLL}")
+  install(FILES ${FDK_AAC_DLL} DESTINATION ${MIXXX_INSTALL_BINDIR})
+elseif(FDK_AAC_LIBRARY)
+  message(STATUS "Found fdk-aac: ${FDK_AAC_LIBRARY}")
```
This message is pointless.

Be-ing replied:
What? No it isn't. It has been really helpful debugging this on CI.

daschuer commented on CMakeLists.txt:
```diff
@@ -2219,29 +2217,49 @@ if(FAAD)
+elseif(FDK_AAC_LIBRARY)
+  message(STATUS "Found fdk-aac: ${FDK_AAC_LIBRARY}")
+else()
+  message(STATUS "Could NOT find fdk-aac")
```
This should be rephrased and give more target specific hints.

Be-ing replied:
Such as?

daschuer replied:
Such as: "Could not find fdk-aac.dll in ${CMAKE_INSTALL_BINDIR}"

Be-ing replied:
I don't think that extra complexity is necessary or really helpful...

JoergAtGithub replied:
Why not just add it? The fast merge of this PR is so important - please don't postpone it uneccessary by such detail discussions!

Be-ing replied:
I agree that merging this PR should not be slowed down by bikeshedding pointless details.

daschuer commented on CMakeLists.txt:
```diff
@@ -2219,29 +2217,49 @@ if(FAAD)
 # FFmpeg is multimedia library that can be found http://ffmpeg.org/
 find_package(FFMPEG COMPONENTS libavcodec libavformat libavutil libswresample)
-option(FFMPEG "FFmpeg 4.x support" OFF)
+cmake_dependent_option(FFMPEG "FFmpeg 4.x support" ON "FFMPEG_FOUND" OFF)
```
Please revert this. It will fail if the FFMPEG version is too old like on Ubuntu Bionic.

Be-ing replied:
I explicitly disabled FFMPEG on Ubuntu 18.04 for this reason.

daschuer replied:
This does not help for home builds. The old version was just fine.

Be-ing replied:
Is anyone actually building on Ubuntu 18.04...

Be-ing replied:
If they are, they will get an error message informing them the version of FFMPEG is too old so they should be able to figure out to disable FFMPEG.

JoergAtGithub replied:
For 2.3 it's still an official supported platform - it must compile out of the box on any supported platform.

Be-ing replied:
FFmpeg is now off by default if it is found but too old a67122042b26aebf33b810e05d27e4e21192f022

daschuer reviewed (dismissed):
I am not comfortable to merge this beast at once. There are many changes, unrelated to vcpkg integration and affecting other platform than only Windows covered by a 51 K lines PR. 
It would be nice to have separate PRs to keep this review-able.

Be-ing commented:
I'm not comfortable breaking this into smaller pieces and leaving the 2.3 branch in an untested, probably broken state.

Be-ing commented:
ping @uklotzde @Holzhaus

Be-ing committed:
CMake: disable FFMPEG automatically if it is too old

Be-ing commented:
Now with TagLib 1.12 on Windows and macOS

uklotzde reviewed (commented):
Just one minor remark where I think an additional comment would help.

The complexity and sheer amount of code is not caused by this PR that just tries to tie everything together. Even if it might (or almost certainly will) cause unforeseeable issues that we missed during the review, I expect those will be fixable. We urgently need some real world tests of this code.

First I was skeptical about switching to the idjc for. But now I consider it the lesser evil compared to the original libshout.

I trust you on the statement that the dynamic loading works on all platforms.

Be-ing committed:
CMake: add comment why qt_de.dm is searched for

Be-ing commented:
Merge?

uklotzde reviewed (approved):
@Holzhaus @daschuer Do you agree?

I would like to merge this as a baseline and resolve any open issues that might occur later. I expect that reviewing individual improvements would be easier and hopefully less controversial afterwards.

Be-ing commented:
The longer this sits open, the more will pile on top.

Be-ing commented:
ping

Be-ing commented:
Is someone going to press merge or are we going to delay this release indefinitely?
